### ISSUE TITLE
[Backend] Op.iter_kernels omits independently implemented external backend Kernels

### ISSUE THREAD

**Issue by lqf0624** [tile-ai/TileOPs#2166]
## Summary

An official InstanceNormFwdOp call successfully executes through our external MUSA builder, but `list(op.iter_kernels())` returns an empty list afterward. The retained upstream cache-reuse test consequently fails.

Our backend owns independent Kernel classes; they do not inherit `tileops.kernels.Kernel`. The enumeration implementation checks that concrete upstream type.

Related: #2164 concerns failure before a normal external builder call. Here, the builder and device execution have already succeeded.

## Tested environment

- TileOPs: `38b7cdc9ec85505eadbc800685e9468cdac54a71`.
- MTT S5000; PyTorch and torch-MUSA `2.9.1.post1+musa5.2.0`.
- TileLang-MUSA `0.1.13+musa.git489df49a`.
- Ordinary wheel installation of an external MUSA backend with an independent InstanceNorm Kernel and registered builder.
- No patches to official Ops or CUDA/MUSA tensor identity.
- Findings apply to this pinned revision; newer main was not device-tested for this issue.

## Upstream code locations

- [Op.iter_kernels, lines 509–521](https://github.com/tile-ai/TileOPs/blob/38b7cdc9ec85505eadbc800685e9468cdac54a71/src/tileops/ops/op_base.py#L509-L521), documented to enumerate every Kernel held by the Op.
- [Op._entry_kernels, lines 535–553](https://github.com/tile-ai/TileOPs/blob/38b7cdc9ec85505eadbc800685e9468cdac54a71/src/tileops/ops/op_base.py#L535-L553), whose direct recognition is:

```python
if isinstance(entry, Kernel):
    return [entry]
```

The other branches walk tuples/lists and dataclass fields. An independent external Kernel object that is not an instance of the upstream Kernel class is otherwise omitted.

- [test_instance_norm_lazy_cache_reuse_and_respecialization, lines 247–277](https://github.com/tile-ai/TileOPs/blob/38b7cdc9ec85505eadbc800685e9468cdac54a71/tests/ops/test_instance_norm.py#L247-L277), expecting one enumerated Kernel after the first specialization, reuse for an identical call, and two after a changed specialization.

## Device reproduction

With the external backend installed and registered:

```python
import torch
import torch_musa
import tileops_musa  # our external backend plugin
from tileops.ops.norm.instance_norm import InstanceNormFwdOp

op = InstanceNormFwdOp()
x = torch.randn(2, 8, 4, 4, dtype=torch.float16, device="musa")
weight = torch.randn(8, dtype=x.dtype, device=x.device)
bias = torch.randn(8, dtype=x.dtype, device=x.device)
y = op(x, weight=weight, bias=bias)  # device execution succeeds
print(len(list(op.iter_kernels())))  # observed: 0
```

The upstream test first compares the output against its FP32 InstanceNorm reference (cast back to the input dtype); that numerical assertion passes. The next assertion fails:

```text
assert len(list(op.iter_kernels())) == 1
E assert 0 == 1
```

## Observed impact and scope

The external implementation is invisible to the enumeration API, preventing the retained test from observing its specialization cache via that API. This failure alone does **not** establish that the underlying cache fails to reuse a Kernel.

The _entry_kernels docstring also notes that entries hidden from this walk are invisible to autotune. We have not independently measured an autotune failure caused by this enumeration issue; the directly reproduced failure here is enumeration after a successful InstanceNorm call.

**Comment by lcy-seso** [comment]
Confirmed on `38b7cdc9`; the relevant code is logically unchanged on main (`252a1721`). The cause is not specific to InstanceNorm. It follows from the backend contract and affects every op served by an external target.

**Cause**

- `BuildKernel` promises only a callable (`backend/protocol.py`), and `Op._build_external` checks only `callable()`.
- The built entry is stored as-is in the role cache. `Op._entry_kernels` recognises only `tileops.kernels.Kernel` instances, sequences, and dataclasses, so `iter_kernels()` skips an independent callable. An external kernel that subclasses `Kernel` is enumerated.

**Impact beyond the reported test**

- `Op.run_config()` returns `None`, so `benchmarks/report.py` reports no config for external runs.
- `Op.autotune()` has no effect on an external target. It sets `op.tune`, but the builder never receives that flag and the protocol defines no autotune hook. Skipping the callable is nonetheless correct: if `_entry_kernels` accepted any callable, `autotune()` would call a `.autotune()` method the callable does not have.
- The `iter_kernels` docstring and `ops-design-reference.md` both describe it as returning every kernel the op holds, which is false under an external target.

**Tests that fail the same way when an external target serves the op**

| Kind | Sites on main |
| --- | --- |
| Cache count after a real call | instance_norm (2 tests; the issue names one), group_norm (3), pool (3), convolution, cumulative, moe_fused_topk, gqa (reuse across sequence lengths), `test_multi_dtype_instance._assert_two_entries` (4 tests), batch_norm lazy cache (3). The batch_norm tests pass `kernel_map=`, but `_resolve_builder` does not consult overrides, so the target still takes over. |
| In-tree kernel type or attributes | Through `iter_kernels`: gqa (type checks), argreduce (`strategy`), kernel_map_install (autotune reach). Through `op.kernel` or `built_kernels(...)`: many more in convolution, gemm, deltanet, cumulative, reduce, moe_grouped_gemm, gqa_prefill_paged. |
| Not affected | softmax (asserts 0 before any build), `tests/test_op_base.py` (fake `Kernel` subclasses, no device). |

**Related gaps in the callable-only protocol**

- `RoPEFwdOp.forward` calls `self.kernel.take_out_of_range()`, which a plain callable does not provide.
- `IndexedRoutedExpert` unpacks four kernels from one entry, and GQA backward stores a two-kernel tuple. `_build_external` rejects a non-callable tuple, so the current protocol cannot serve these ops.

**Fix**

1. `iter_kernels()` will be defined as returning the in-tree `Kernel` instances, which is the set `autotune()` acts on. The docstring and `ops-design-reference.md` will be corrected in the same change.
2. The count tests will be split. In-tree specialization tests will pin `target=BUILTIN`. External cache reuse will get a dedicated test at the builder boundary, with a fake builder that counts its calls. `len(built_kernels(role))` will not be used as a substitute: external entries are keyed on every input's full shape, so the GQA reuse-across-sequence-lengths test would count one entry per shape.
3. Every test that checks kernel type, `kernel_map` overrides, or kernel attributes will pin `target=BUILTIN`.
4. The RoPE and compound-entry protocol gaps will be tracked in a separate issue, alongside #2164.

**Comment by lcy-seso** [comment]
Hi @lqf0624, thank you for this detailed report! This should be fixed on the latest main. I've reopened the issue so you can check whether the fix works in your environment. If you still see a problem, please feel free to open a new issue, and we'll look at it quickly. Thanks again!
